src/redisearch_rs/slots_tracker/src/slots_tracker.rs

```diff
@@ -0,0 +1,931 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Safe implementation of the slots tracker state.
+//!
+//! This module provides a single-thread-safe implementation of the slots tracker.
+//! All methods take `&mut self`, making the borrowing rules enforce single-threaded access.
+
+use std::num::NonZeroU32;
+
+use crate::SlotRange;
+use crate::slot_set::{CoverageRelation, SlotSet};
+
+/// Represents the version state of slot configuration.
+///
+/// This enum encapsulates the version logic and avoids exposing magic numbers.
+/// The discriminant itself IS the version value, making this exactly 32 bits.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Version {
+    /// Stable version values: 1..=u32::MAX
+    /// Each value represents a specific version counter.
+    Stable(NonZeroU32),
+    /// Unstable state marker
+    /// Slots are available but configuration is changing.
+    Unstable,
+}
+const _: () = assert!(std::mem::size_of::<Version>() == 4);
+
+impl Default for Version {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Version {
+    /// Creates a new Version initialized to version 1 (stable).
+    const fn new() -> Self {
+        // SAFETY: value is not zero
+        Self::Stable(unsafe { NonZeroU32::new_unchecked(1) })
+    }
+
+    /// Increments a stable version using wrapping arithmetic.
+    ///
+    /// # Panics
+    ///
+    /// Panics in debug mode if called on an Unstable version.
+    fn increment(self) -> Self {
+        match self {
+            Self::Stable(v) => {
+                if v.get() < u32::MAX {
+                    // SAFETY: 1 < v + 1 <= u32::MAX, so not zero
+                    Self::Stable(unsafe { NonZeroU32::new_unchecked(v.get() + 1) })
+                } else {
+                    // Wrap around to 1 from u32::MAX
+                    // SAFETY: value is not zero
+                    Self::Stable(unsafe { NonZeroU32::new_unchecked(1) })
+                }
```

Comment:
SInce you are doing all this checking anyway, why not use a normal `u32` for the `Stable` variant? Mostly to remove the need for this `unsafe` code

Reply:
It makes the `Version` size 64bit.
I can avoid the unsafe new with `NonZeroU32::new(1).unwrap()`.
Or make Stable use u16